File: sdk/go/sharing.go
package psnx

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"math/big"
)

// Prime for GF(p) - 2^256 - 189
var prime *big.Int

func init() {
	prime = new(big.Int).Exp(big.NewInt(2), big.NewInt(256), nil)
	prime.Sub(prime, big.NewInt(189))
}

// Share represents a Shamir secret share
type Share struct {
	Index     int    `json:"index"`
	Data      []byte `json:"data"`
	Threshold int    `json:"threshold"`
	Total     int    `json:"total"`
	Checksum  string `json:"checksum"`
}

// SecretSharing implements Shamir's Secret Sharing
type SecretSharing struct {
	threshold int
	total     int
}

// NewSecretSharing creates a new secret sharing instance
func NewSecretSharing(threshold, total int) (*SecretSharing, error) {
	if threshold < 2 {
		return nil, errors.New("threshold must be >= 2")
	}
	if total < threshold {
		return nil, errors.New("total must be >= threshold")
	}
	if total > 255 {
		return nil, errors.New("maximum 255 shares")
	}

	return &SecretSharing{
		threshold: threshold,
		total:     total,
	}, nil
}

// Split divides a secret into N shares
func (ss *SecretSharing) Split(secret []byte) ([]*Share, error) {
	if len(secret) > 32 {
		return nil, errors.New("secret too large (max 32 bytes)")
	}

	// Convert secret to big int
	secretInt := new(big.Int).SetBytes(secret)

	// Generate polynomial coefficients
	coefficients := make([]*big.Int, ss.threshold)
	coefficients[0] = secretInt

	for i := 1; i < ss.threshold; i++ {
		randBytes := make([]byte, 32)
		if _, err := rand.Read(randBytes); err != nil {
			return nil, err
		}
		coeff := new(big.Int).SetBytes(randBytes)
		coeff.Mod(coeff, new(big.Int).Sub(prime, big.NewInt(1)))
		coeff.Add(coeff, big.NewInt(1))
		coefficients[i] = coeff
	}

	// Evaluate polynomial at N points
	shares := make([]*Share, ss.total)
	for x := 1; x <= ss.total; x++ {
		y := ss.evaluate(coefficients, big.NewInt(int64(x)))

		// Convert to 32 bytes
		yBytes := make([]byte, 32)
		yBytesRaw := y.Bytes()
		copy(yBytes[32-len(yBytesRaw):], yBytesRaw)

		// Calculate checksum
		indexBytes := make([]byte, 4)
		binary.BigEndian.PutUint32(indexBytes, uint32(x))
		toHash := append(yBytes, indexBytes...)
		hash := sha256.Sum256(toHash)
		checksum := bytesToHex(hash[:4])

		shares[x-1] = &Share{
			Index:     x,
			Data:      yBytes,
			Threshold: ss.threshold,
			Total:     ss.total,
			Checksum:  checksum,
		}
	}

	return shares, nil
}

// Reconstruct recovers the secret from K shares
func (ss *SecretSharing) Reconstruct(shares []*Share) ([]byte, error) {
	if len(shares) < ss.threshold {
		return nil, errors.New("not enough shares")
	}

	// Verify checksums
	for _, share := range shares {
		if !ss.VerifyShare(share) {
			return nil, errors.New("share corrupted: " + string(rune(share.Index)))
		}
	}

	// Use first K shares
	useShares := shares[:ss.threshold]

	// Check for duplicates
	seen := make(map[int]bool)
	for _, s := range useShares {
		if seen[s.Index] {
			return nil, errors.New("duplicate shares")
		}
		seen[s.Index] = true
	}

	// Convert to points
	points := make([][2]*big.Int, len(useShares))
	for i, s := range useShares {
		points[i] = [2]*big.Int{
			big.NewInt(int64(s.Index)),
			new(big.Int).SetBytes(s.Data),
		}
	}

	// Lagrange interpolation at x=0
	secretInt := ss.lagrange(points, big.NewInt(0))

	// Convert back to bytes
	secretBytes := secretInt.Bytes()

	return secretBytes, nil
}

func (ss *SecretSharing) evaluate(coefficients []*big.Int, x *big.Int) *big.Int {
	result := big.NewInt(0)
	xPower := big.NewInt(1)

	for _, coeff := range coefficients {
		term := new(big.Int).Mul(coeff, xPower)
		result.Add(result, term)
		result.Mod(result, prime)
		xPower.Mul(xPower, x)
		xPower.Mod(xPower, prime)
	}

	return result
}

func (ss *SecretSharing) lagrange(points [][2]*big.Int, x *big.Int) *big.Int {
	result := big.NewInt(0)

	for i, point := range points {
		xi, yi := point[0], point[1]

		num := big.NewInt(1)
		den := big.NewInt(1)

		for j, other := range points {
			if i != j {
				xj := other[0]

				// num *= (x - xj)
				diff := new(big.Int).Sub(x, xj)
				num.Mul(num, diff)
				num.Mod(num, prime)

				// den *= (xi - xj)
				diff = new(big.Int).Sub(xi, xj)
				den.Mul(den, diff)
				den.Mod(den, prime)
			}
		}

		// Modular inverse of den
		denInv := new(big.Int).ModInverse(den, prime)
		if denInv == nil {
			continue
		}

		// coeff = num * denInv
		coeff := new(big.Int).Mul(num, denInv)
		coeff.Mod(coeff, prime)

		// result += yi * coeff
		term := new(big.Int).Mul(yi, coeff)
		result.Add(result, term)
		result.Mod(result, prime)
	}

	// Handle negative result
	if result.Sign() < 0 {
		result.Add(result, prime)
	}

	return result
}

// VerifyShare checks the checksum of a share
func (ss *SecretSharing) VerifyShare(share *Share) bool {
	indexBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(indexBytes, uint32(share.Index))
	toHash := append(share.Data, indexBytes...)
	hash := sha256.Sum256(toHash)
	expected := bytesToHex(hash[:4])
	return share.Checksum == expected
}

File: sdk/rust/src/sharing.rs
//! Shamir Secret Sharing

use num_bigint::BigUint;
use num_traits::{One, Zero};
use rand::{rngs::OsRng, RngCore};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::{Error, Result};

/// Prime for GF(p)
fn prime() -> BigUint {
    BigUint::from(2u32).pow(256) - BigUint::from(189u32)
}

/// A share of the secret
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Share {
    pub index: u8,
    pub data: Vec<u8>,
    pub threshold: u8,
    pub total: u8,
    pub checksum: String,
}

impl Share {
    /// Verify the checksum
    pub fn verify(&self) -> bool {
        let mut hasher = Sha256::new();
        hasher.update(&self.data);
        hasher.update(&[self.index]);
        let hash = hasher.finalize();
        let expected = hex::encode(&hash[..4]);
        self.checksum == expected
    }
    
    /// To base64 JSON
    pub fn to_json(&self) -> Result<String> {
        Ok(serde_json::to_string(&ShareB64 {
            index: self.index,
            data: base64::encode(&self.data),
            threshold: self.threshold,
            total: self.total,
            checksum: self.checksum.clone(),
        })?)
    }
    
    /// From base64 JSON
    pub fn from_json(json: &str) -> Result<Self> {
        let b64: ShareB64 = serde_json::from_str(json)?;
        Ok(Self {
            index: b64.index,
            data: base64::decode(&b64.data)?,
            threshold: b64.threshold,
            total: b64.total,
            checksum: b64.checksum,
        })
    }
}

#[derive(Serialize, Deserialize)]
struct ShareB64 {
    index: u8,
    data: String,
    threshold: u8,
    total: u8,
    checksum: String,
}

/// Shamir Secret Sharing scheme
pub struct SecretSharing {
    k: u8, // threshold
    n: u8, // total
}

impl SecretSharing {
    /// Create a new sharing scheme
    pub fn new(threshold: u8, total: u8) -> Result<Self> {
        if threshold < 2 {
            return Err(Error::Validation("Threshold must be >= 2".to_string()));
        }
        if total < threshold {
            return Err(Error::Validation("Total must be >= threshold".to_string()));
        }
        
        Ok(Self { k: threshold, n: total })
    }
    
    /// Split secret into N shares
    pub fn split(&self, secret: &[u8]) -> Result<Vec<Share>> {
        if secret.len() > 32 {
            return Err(Error::Validation("Secret too large (max 32 bytes)".to_string()));
        }
        
        let p = prime();
        let secret_int = BigUint::from_bytes_be(secret);
        
        // Generate polynomial coefficients
        let mut coefficients = vec![secret_int];
        for _ in 1..self.k {
            let mut rand_bytes = vec![0u8; 32];
            OsRng.fill_bytes(&mut rand_bytes);
            let coeff = BigUint::from_bytes_be(&rand_bytes) % &p;
            coefficients.push(if coeff.is_zero() { BigUint::one() } else { coeff });
        }
        
        // Evaluate at N points
        let mut shares = Vec::with_capacity(self.n as usize);
        for x in 1..=self.n {
            let x_big = BigUint::from(x);
            let y = self.evaluate(&coefficients, &x_big, &p);
            
            // Pad to 32 bytes
            let mut y_bytes = y.to_bytes_be();
            while y_bytes.len() < 32 {
                y_bytes.insert(0, 0);
            }
            
            // Checksum
            let mut hasher = Sha256::new();
            hasher.update(&y_bytes);
            hasher.update(&[x]);
            let hash = hasher.finalize();
            let checksum = hex::encode(&hash[..4]);
            
            shares.push(Share {
                index: x,
                data: y_bytes,
                threshold: self.k,
                total: self.n,
                checksum,
            });
        }
        
        Ok(shares)
    }
    
    /// Reconstruct secret from K shares
    pub fn reconstruct(&self, shares: &[Share]) -> Result<Vec<u8>> {
        if shares.len() < self.k as usize {
            return Err(Error::Share(format!(
                "Need {} shares, got {}",
                self.k,
                shares.len()
            )));
        }
        
        // Verify checksums
        for share in shares {
            if !share.verify() {
                return Err(Error::Share(format!("Share {} corrupted", share.index)));
            }
        }
        
        let p = prime();
        
        // Use first K shares
        let use_shares: Vec<_> = shares.iter().take(self.k as usize).collect();
        
        // Check uniqueness
        let mut indices: Vec<_> = use_shares.iter().map(|s| s.index).collect();
        indices.sort();
        indices.dedup();
        if indices.len() != self.k as usize {
            return Err(Error::Share("Duplicate shares".to_string()));
        }
        
        // Points (x, y)
        let points: Vec<_> = use_shares
            .iter()
            .map(|s| (BigUint::from(s.index), BigUint::from_bytes_be(&s.data)))
            .collect();
        
        // Lagrange interpolation
        let secret_int = self.lagrange(&points, &BigUint::zero(), &p);
        
        // Convert to bytes
        let mut bytes = secret_int.to_bytes_be();
        
        // Remove leading zeros but keep at least one byte
        while bytes.len() > 1 && bytes[0] == 0 {
            bytes.remove(0);
        }
        
        Ok(bytes)
    }
    
    fn evaluate(&self, coefficients: &[BigUint], x: &BigUint, p: &BigUint) -> BigUint {
        let mut result = BigUint::zero();
        let mut x_power = BigUint::one();
        
        for coeff in coefficients {
            result = (&result + coeff * &x_power) % p;
            x_power = (&x_power * x) % p;
        }
        
        result
    }
    
    fn lagrange(&self, points: &[(BigUint, BigUint)], x: &BigUint, p: &BigUint) -> BigUint {
        let mut result = BigUint::zero();
        
        for (i, (xi, yi)) in points.iter().enumerate() {
            let mut num = BigUint::one();
            let mut den = BigUint::one();
            
            for (j, (xj, _)) in points.iter().enumerate() {
                if i != j {
                    // num = num * (x - xj)
                    let diff = if x >= xj {
                        x - xj
                    } else {
                        p - (xj - x) % p
                    };
                    num = (&num * &diff) % p;
                    
                    // den = den * (xi - xj)
                    let diff = if xi >= xj {
                        xi - xj
                    } else {
                        p - (xj - xi) % p
                    };
                    den = (&den * &diff) % p;
                }
            }
            
            // Modular inverse
            let den_inv = self.mod_inverse(&den, p);
            let coeff = (&num * &den_inv) % p;
            result = (&result + yi * &coeff) % p;
        }
        
        result
    }
    
    fn mod_inverse(&self, a: &BigUint, m: &BigUint) -> BigUint {
        // Extended Euclidean algorithm using Fermat's little theorem
        // a^(-1) = a^(p-2) mod p for prime p
        let exp = m - BigUint::from(2u32);
        a.modpow(&exp, m)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_split_reconstruct() {
        let ss = SecretSharing::new(3, 5).unwrap();
        let secret = b"my secret key!!!"; // 16 bytes
        
        let shares = ss.split(secret).unwrap();
        assert_eq!(shares.len(), 5);
        
        // Reconstruct with 3 shares
        let recovered = ss.reconstruct(&shares[0..3]).unwrap();
        assert_eq!(secret.to_vec(), recovered);
    }
}
